src/Data/Path/Pathy.purs

```diff
@@ -478,6 +476,18 @@ module Data.Path.Pathy
   parseAbsDir :: String -> Maybe (AbsDir Unsandboxed)
   parseAbsDir = parsePath (const Nothing) (const Nothing) (const Nothing) Just
 
+  -- | Folds a `Path` into a value using two specified functions and a provided
+  -- | default value.
+  -- |
+  -- | The first function folds a `FileName` into our value, and the second
+  -- | function folds a `DirName` into our value.
+  fold :: forall c a b s. (FileName -> c -> c) -> (DirName -> c -> c) -> c -> Path a b s -> c
+  fold f g z (Current      ) = z
+  fold f g z (Root         ) = z
+  fold f g z (ParentIn p   ) = fold f g z p
```

Comment:
This relationship is interesting and should probably be exposed as part of the fold.

When I was last playing around with adding this, I had something like this:

```
data NamelessNode = ParentPath | CurrentPath | RootPath

-- | Folds over a path, starting at the base of the path and working outward.
foldPath :: forall a b s z. (z -> NamelessNode -> z) -> (z -> Either (Tuple (Path a Dir s) DirName) (Tuple (Path a Dir s) FileName) -> z) -> z -> Path a b s -> z
```

I can speak to how useful this is, but my goal was in *precisely* capturing the structure of the path (including parent in relationships). In fact, with such a fold function, I wanted to replace the remaining "unsafe" implementations (which switch type parameters around in an unchecked fashion).